### ISSUE TITLE
stringstream wrongly mapped to iostreams

### ISSUE THREAD

**Issue by Flamefire** [include-what-you-use/include-what-you-use#277]
Check the following example: 

```
#include <iostream>
#include <sstream>

int main(){
  std::stringstream s;
  s << 1;
  std::cout << s;
}
```

iwyu suggest removing `#include <sstream>` which is of course wrong. The info shows, that it maps this to `iostream`

**Comment by vsapsai** [comment]
Thanks for the report, @Flamefire. Issue [#90: fstream include removed while required](https://github.com/include-what-you-use/include-what-you-use/issues/90) is very similar. And there is some iosfwd discussion in issue [#80: Test badinc fails on Mac](https://github.com/include-what-you-use/include-what-you-use/issues/80).

**Comment by Flamefire** [comment]
From my analysis the problem is at the point where it says something like: "private headers export, what their public headers export". This breaks, as "private" is not really "private" at that point. You have a mapping from `<iosfwd>` to `<sstream>` so it adds everything exported from `sstream` to `iosfwd`. This should be done only (if at all) if the latter is actually a private header and not just a mapping

**Comment by firewave** [comment]
```cpp
#include <iostream>
#include <sstream>

int main(void)
{
    std::istringstream istr;
}
```

```
input.cpp should add these lines:

input.cpp should remove these lines:
- #include <sstream>  // lines 2-2

The full include-list for input.cpp:
#include <iostream>  // for istringstream
---
```

```cpp
#include <iostream>
#include <sstream>
#include <iosfwd>

int main(void)
{
    std::istringstream istr;
}
```

```
input.cpp should add these lines:

input.cpp should remove these lines:
- #include <iostream>  // lines 1-1
- #include <sstream>  // lines 2-2

The full include-list for input.cpp:
#include <iosfwd>  // for istringstream
---
```

**Comment by nealsid** [comment]
I am running into this on IWYU 17/Clang 13, as well.  Thanks!

**Comment by Oscardoras** [comment]
I have a similar issue with IWYU 0.20/Clang 16.

ChannelID.h:
`
#pragma once

#include <ostream>
#include <string>

class ChannelID final {

  std::string toString() const;

  friend std::ostream& operator<<( std::ostream& str, const ChannelID& obj );

};
`

ChannelID.cpp:
`
#include "ChannelID.h"
#include <sstream>

std::string ChannelID::toString() const {
  std::ostringstream o;
  return o.str();
}

std::ostream& operator<<( std::ostream& s, const ChannelID& obj ) {
  return s;
}
`

And I get:
`
Warning: include-what-you-use reported diagnostics:

(ChannelID.h has correct #includes/fwd-decls)

ChannelID.cpp should add these lines:

ChannelID.cpp should remove these lines:
- #include <sstream>  // lines 2-2

The full include-list for ChannelID.cpp:
#include "ChannelID.h"
---
`

And this issue appears because of the ostream in ChannelID.h.

**Comment by firewave** [comment]
It appears this has been resolved with 0.21. Using `include-what-you-use-0.21-1.1` on OpenSUSE Tumbleweed I was able to remove all the `IWYU pragma: keep` for `sstream` and `fstream` includes.

Unfortunately I do not have that version on a local installation so I cannot completely verify this.

**Comment by kimgr** [comment]
This was silently fixed in https://github.com/include-what-you-use/include-what-you-use/commit/bc5221b570f4c70eb2601a79c79da2832a1c7aac, released as part of IWYU 0.21.

Thanks @bolshakov-a for taking care of this!
